src/typechecker.erl

```diff
@@ -3499,52 +3499,69 @@ type_check_function(Env, {function,_, Name, NArgs, Clauses}) ->
+-spec do_add_types_pats(Pats :: [gradualizer_type:abstract_pattern()],
+                        Tys  :: [type()],
+                        TEnv :: #tenv{},
+                        VEnv :: map()
+                       ) -> {PatTys      :: [type()],
+                             UBounds     :: [type()],
+                             NewVEnv     :: map(),
+                             Constraints :: constraints:constraints()}.
+do_add_types_pats(Pats, Tys, TEnv, VEnv) ->
+    add_types_pats(Pats, Tys, TEnv, VEnv, [], [], []).
+
+add_types_pats([], [], _TEnv, VEnv, PatTysAcc, UBoundsAcc, CsAcc) ->
     {lists:reverse(PatTysAcc), lists:reverse(UBoundsAcc),
      VEnv, constraints:combine(CsAcc)};
-add_types_pats([Pat | Pats], [Ty | Tys], TEnv, VEnv, Caps, PatTysAcc, UBoundsAcc, CsAcc) ->
+add_types_pats([Pat | Pats], [Ty | Tys], TEnv, VEnv, PatTysAcc, UBoundsAcc, CsAcc) ->
     NormTy = normalize(Ty, TEnv),
     {PatTyNorm, UBoundNorm, VEnv2, Cs1} =
-        ?throw_orig_type(add_type_pat(Pat, NormTy, TEnv, VEnv, Caps),
+        ?throw_orig_type(add_type_pat(Pat, NormTy, TEnv, VEnv),
                          Ty, NormTy),
     %% De-normalize the returned types if they are the type checked against.
     PatTy  = case PatTyNorm  of NormTy -> Ty;
                                 _      -> PatTyNorm end,
     UBound = case UBoundNorm of NormTy -> Ty;
                                 _      -> UBoundNorm end,
-    add_types_pats(Pats, Tys, TEnv, VEnv2, Caps,
+    add_types_pats(Pats, Tys, TEnv, VEnv2,
                    [PatTy|PatTysAcc], [UBound|UBoundsAcc], [Cs1|CsAcc]).
 
 %% Type check a pattern against a normalized type and add variable bindings.
+%% NB: For correct variable binding and refinement logic, don't use this
+%% function directly.  Instead, use add_types_pats([Pat], [Type], TEnv, VEnv).
 -spec add_type_pat(Pat  :: gradualizer_type:abstract_pattern(),
                    Type :: type(),
                    TEnv :: #tenv{},
-                   VEnv :: map(),
-		   Caps :: capture_vars | bind_vars) ->
+                   VEnv :: map()) ->
           {PatTy :: type(), UBound :: type(), NewVEnv :: map(),
            constraints:constraints()}.
-add_type_pat({var, _, '_'}, Ty, _TEnv, VEnv, _) ->
+add_type_pat({var, _, '_'}, Ty, _TEnv, VEnv) ->
     {Ty, Ty, VEnv, constraints:empty()};
-add_type_pat({var, _, A} = Var, Ty, TEnv, VEnv, CapOrBind) ->
+add_type_pat({var, _, A} = Var, Ty, TEnv, VEnv) ->
     case VEnv of
-	#{A := VarTy} when CapOrBind == capture_vars ->
-	    case subtype(VarTy, Ty, TEnv) of
-		{true, Cs} ->
-		    {type(none), VarTy, VEnv, Cs};
-		false ->
-		    throw({type_error, Var, VarTy, Ty})
+	#{A := VarTy} ->
+	    case glb(VarTy, Ty, TEnv) of
+                {?type(none), _Cs} ->
+                    %% TODO: Better type error (it's a pattern, not an
+                    %% expression)
+		    throw({type_error, Var, VarTy, Ty});
+		{RefinedTy, Cs} ->
+		    {type(none), RefinedTy, VEnv#{A := RefinedTy}, Cs}
```

Comment:
It's nice to be able to get rid of the `capture_vars` logic here.

Reply:
:+1: